File: apps/haskell/src/Ch04/ThreadBasics.hs
module Ch04.ThreadBasics
    ( createThread
    , runThreads
    ) where

import Control.Concurrent (forkIO, ThreadId)
import Control.Concurrent.Async (async, wait, Async)
import Control.Monad (forM, forM_)

-- | Create and start a new thread
createThread :: IO () -> IO ThreadId
createThread = forkIO

-- | Run multiple threads and wait for completion
runThreads :: Int -> (Int -> IO ()) -> IO ()
runThreads count action = do
    handles <- forM [0..count-1] $ \i -> async (action i)
    forM_ handles wait

File: apps/scala/src/main/scala/concurrency/ch04/ThreadBasics.scala
package concurrency.ch04

object ThreadBasics:

  /** Create a worker thread with the given name and task */
  def createWorker(name: String, task: () => Unit): Thread =
    val thread = new Thread(() => task())
    thread.setName(name)
    thread

  /** Create multiple worker threads */
  def createMultipleWorkers(count: Int, task: Int => Unit): List[Thread] =
    (0 until count).map { i =>
      val thread = new Thread(() => task(i))
      thread.setName(s"Worker-$i")
      thread
    }.toList

  /** Start all threads and wait for them to complete */
  def startAndJoinAll(threads: List[Thread]): Unit =
    threads.foreach(_.start())
    threads.foreach(_.join())
